packages/adapter-cloudflare-workers/files/entry.js

```diff
@@ -1,57 +1,96 @@
+const static_asset_manifest = JSON.parse(static_asset_manifest_json);
 
 const server = new Server(manifest);
 
 const prefix = `/${manifest.appDir}/`;
 
-addEventListener('fetch', (/** @type {FetchEvent} */ event) => {
-	event.respondWith(handle(event));
-});
+export default {
+	/**
+	 * @param {Request} req
+	 * @param {any} env
+	 * @param {any} context
+	 */
+	async fetch(req, env, context) {
+		const url = new URL(req.url);
 
-/**
- * @param {FetchEvent} event
- * @returns {Promise<Response>}
- */
-async function handle(event) {
-	const { request } = event;
-
-	const url = new URL(request.url);
-
-	// generated assets
-	if (url.pathname.startsWith(prefix)) {
-		const res = await getAssetFromKV(event);
-		return new Response(res.body, {
-			headers: {
-				'cache-control': 'public, immutable, max-age=31536000',
-				'content-type': res.headers.get('content-type')
+		// static assets
+		if (url.pathname.startsWith(prefix)) {
+			/** @type {Response} */
+			const res = await get_asset_from_kv(req, env, context);
+			if (is_error(res.status)) {
+				return res;
 			}
-		});
-	}
+			return new Response(res.body, {
+				headers: {
+					// include original cache headers, minus cache-control which
+					// is overridden, and etag which is no longer useful
+					'cache-control': 'public, immutable, max-age=31536000',
+					'content-type': res.headers.get('content-type'),
+					'x-robots-tag': 'noindex'
+				}
+			});
+		}
 
-	// prerendered pages and index.html files
-	const pathname = url.pathname.replace(/\/$/, '');
-	let file = pathname.substring(1);
+		// prerendered pages and index.html files
+		const pathname = url.pathname.replace(/\/$/, '');
+		let file = pathname.substring(1);
 
-	try {
-		file = decodeURIComponent(file);
-	} catch (err) {
-		// ignore
-	}
+		try {
+			file = decodeURIComponent(file);
+		} catch (err) {
+			// ignore
+		}
+
+		if (
+			manifest.assets.has(file) ||
+			manifest.assets.has(file + '/index.html') ||
+			prerendered.has(pathname || '/')
+		) {
+			return get_asset_from_kv(req, env, context);
+		}
 
-	if (
-		manifest.assets.has(file) ||
-		manifest.assets.has(file + '/index.html') ||
-		prerendered.has(pathname || '/')
-	) {
-		return await getAssetFromKV(event);
+		// dynamically-generated pages
+		try {
+			return await server.respond(req, { platform: { env, context } });
+		} catch (e) {
+			return new Response('Error rendering route: ' + (e.message || e.toString()), { status: 500 });
+		}
 	}
+};
 
-	// dynamically-generated pages
+/**
+ * @param {Request} req
+ * @param {any} env
+ * @param {any} context
+ */
+async function get_asset_from_kv(req, env, context) {
 	try {
-		return await server.respond(request);
+		return await getAssetFromKV(
+			{
+				request: req,
+				waitUntil(promise) {
+					return context.waitUntil(promise);
+				}
+			},
+			{
+				ASSET_NAMESPACE: env.__STATIC_CONTENT,
+				ASSET_MANIFEST: static_asset_manifest
+			}
```

Comment:
mentioned this in discord, but for posterity: would be good to understand why (if) this is necessary

Reply:
taken directly from the docs of kv-asset-handler: https://github.com/cloudflare/kv-asset-handler#es-modules

maybe someone from CF or with more experience using it can chime in why it is done that way?

Reply:
> The binding name to the KV Namespace populated with key/value entries of files for the Worker to serve. By default, Workers Sites uses a [binding to a Workers KV Namespace](https://developers.cloudflare.com/workers/reference/storage/api/#namespaces) named __STATIC_CONTENT.

Awkward, but is what it is. Both are required

Reply:
I don't understand how the existing adapter works if this is true? Is it a different between the service worker and module formats? (Also that link is broken btw :)

Reply:
@Rich-Harris Service workers get KV namespaces and env variables bound to global, whereas for module workers, they are provided in an env parameter passed to the exported [`fetch` function](https://github.com/sveltejs/kit/blob/4db18af228ac4960adbfce04ef46fda1cf219f2a/packages/adapter-cloudflare-workers/files/entry.js#L17). The current implementation for service workers doesn't pass any options to `getAssetFromKV`, so it [defaults to using the globally bound `__STATIC_CONTENT`](https://github.com/cloudflare/kv-asset-handler/blob/fd27ccfeb541312f5e6c6be74932efc32ae4ce1f/src/index.ts#L24-L25).
Sorry to butt in, just looking forward to this getting merged ;)

Reply:
> (Also that link is broken btw :)

https://developers.cloudflare.com/workers/runtime-apis/kv/#referencing-kv-from-durable-objects-and-workers-using-modules-syntax I think Luke meant this, but in any case this explains it.

Reply:
Thanks. Those docs suggest we should do this...

```js
const valueFromKV = await this.env.NAMESPACE.get('someKey')
return new Response(valueFromKV)
```

...rather than importing `getAssetFromKV` and passing in `ASSET_NAMESPACE` and `ASSET_MANIFEST`. (Though I don't quite understand how that sets the correct content type etc, and it's not clear from the docs how `someKey` relates to `request.url`) Honestly I'm a bit flummoxed at this point!

Reply:
That specific example is for Durable Objects, so not 100% applicable but consider this code example https://blog.cloudflare.com/workers-javascript-modules/#:~:text=body%3A%20error.stack%0A%20%20%7D)%0A%7D-,//%20filename%3A%20./src/worker.js,-import%20%7B%20logError%20%7D%20from

Reply:
You can do it directly without the `kv-asset-handler` package, but then you have to do the lookups & the content-type setting yourself. (This is what I do, I don't use this pkg). The `0.2.0` of this pkg had an API change to allow for/accommodate Module Worker syntax. Before then, it just assumed & grabbed the namespace binding (aka global variable) that Workers Sites uses. In a Module Worker world, there are no globals and you have to be explicit, hence the 2nd object of extra stuff.

The namespace is to actually do the lookups. 
The manifest is to know what paths exist (or don't) in the namespace. 

The pkg [deals with mime types](https://github.com/cloudflare/kv-asset-handler/blob/master/src/index.ts#L160) as is & does some other footwork for cache management, which we don't really need here imo.

Reply:
> This is what I do, I don't use this pkg

is that in another importable package or can you share it here? Or should we stick to kvAssetHandler regardless to reduce maintenance here. Its 312 lines currently and thats counting the comments. How bad can it be? (I feel i might regret this question later)

Reply:
https://github.com/lukeed/worktop/blob/next/packages/worktop/src/cfw.kv.assets.ts

```ts
import { serve } from 'worktop/cfw.kv.assets';

// ...

let res = await serve(env.__STATIC_CONTENT, request);
return res || new Response('Missing', { status: 404 });
```

Doesn't need the mapping, but you can do some routing with it outside of the `serve` call & then pass down the pre-determined pathname as the 2nd parameter instead of `request`; eg `serve(env.__STATIC_CONTENT, '/foobar.webm')`;

Reply:
Note that we have access to all relevant MIME types in `manifest`, along with a list of the static assets themselves. Is there a way we can use this rather than importing a list of types, and `__STATIC_CONTENT_MANIFEST` for that matter?

Right now the MIME types are on the private `manifest._.mime` object, but we could make it public on `manifest.mime` if it's useful here.

Reply:
Yes, didn't know that existed. If that could be passed down to the handler then, that's all you'd need. No external library, just a KV read and probably grab the `Range` logic from my last link.

No idea what `__STATIC_CONTENT_MANIFEST` is – thought this was SK-specific too.

Reply:
I've opened https://github.com/sveltejs/kit/pull/4302 — if we merge that, then this PR can use `manifest.mime` to construct its own responses.

Presumably we'd want to set ETags as well? Not sure what that looks like in practice

Reply:
Handling cache headers / etags is needed. The current approach of far-future caching hashed assets is good, this PR had just added some error handling to it.

If we were to replace kvAssetHandler with our own logic, do we need to handle edge caching too? What about prerender results? At some point we're ending up replementing a lot of things that are mostly generic. 

Would it be more useful to do this in a reusable/external way, eg in worktop or via sending PRs to kvAssetHandler. Or even having a separate "cloudflare-adapter-utils" library that can be shared between the two cf adapters.

I'm no longer confident we can easily merge the two because of the fundamental difference that one needs a local build step and the other one builds on cf.

Reply:
We should serve prerender results, yeah. As for ETags, we probably want to generate content hashes at build time, which has implications for the shape of `SSRManifest`

Reply:
`__STATIC_CONTENT_MANIFEST` is also necessary because Wrangler adds an additional fingerprint to every asset it uploads, and it contains the mapping from the requested asset to the fingerprinted one.

Reply:
What kind of fingerprint?

Reply:
Here's an example with a test app I deployed:

<img width="382" alt="Screen Shot 2022-04-11 at 10 26 19 AM" src="https://user-images.githubusercontent.com/685518/162761177-10ae39d5-b0b6-4648-a982-879090dde6ad.png">

I'm unsure what adds these extra fingerprints, or why they need to exist, but without the `__STATIC_CONTENT_MANIFEST` the files will not be found.

This is definitely not ideal, but I think it's acceptable given as soon as Pages is out of beta, everyone is going to drop Workers and migrate over. If for some reason that doesn't happen, I'd be happy to dig in further and try to optimize this more, but until that's certain I think it's best to use the standard Cloudflare methodology for the time being.